// @slack/interactive-messages/dist/express-middleware.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.errorCodes = undefined;
exports.createExpressMiddleware = createExpressMiddleware;

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _util = require('./util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var errorCodes = exports.errorCodes = {
  NO_BODY_PARSER: 'SLACKMESSAGEMIDDLEWARE_NO_BODY_PARSER',
  TOKEN_VERIFICATION_FAILURE: 'SLACKMESSAGEMIDDLEWARE_TOKEN_VERIFICATION_FAILURE'
};

var debug = (0, _debug2.default)('@slack/interactive-messages:express-middleware');

function createExpressMiddleware(adapter) {
  var poweredBy = (0, _util.packageIdentifier)();

  // This function binds a specific response instance to a function
  function sendResponse(res) {
    return function _sendResponse(dispatchResult) {
      var status = dispatchResult.status,
          content = dispatchResult.content;

      var contentReady = content && typeof content.then === 'function' ? content : Promise.resolve(content);
      return contentReady.then(function (c) {
        res.status(status);
        res.set('X-Slack-Powered-By', poweredBy);
        if (c) {
          res.json(c);
        } else {
          res.end();
        }
        return new Promise(function (resolve, reject) {
          res.on('finish', function () {
            // res._headers is an undocumented property, but we feel comfortable using it because:
            // 1. express depends on it and express is so foundational in node
            // 2. this is logging code and the risk of this causing a break is minimal
            // eslint-disable-next-line no-underscore-dangle
            debug('response finished - status: %d, headers: %o', res.statusCode, res._headers);
            resolve(res);
          });
          res.on('error', reject);
        });
      });
    };
  }

  return function slackMessageAdapterMiddleware(req, res, next) {
    debug('request received - method: %s, path: %s', req.method, req.path);

    // Bind a response function to this request's respond object. This may be used in a number of
    // places
    var respond = sendResponse(res);

    // Check that the request body has been parsed
    if (!req.body) {
      var error = new Error('The incoming HTTP request did not have a parsed body.');
      error.code = errorCodes.NO_BODY_PARSER;
      next(error);
      return;
    }

    if (req.body.ssl_check) {
      respond({ status: 200 });
      return;
    }

    var payload = JSON.parse(req.body.payload);

    // Handle request token verification
    if (!payload.token || payload.token !== adapter.verificationToken) {
      debug('request token verification failure');
      var _error = new Error('Slack interactive message verification failed');
      _error.code = errorCodes.TOKEN_VERIFICATION_FAILURE;
      next(_error);
      return;
    }
    debug('request token verification success');

    respond(adapter.dispatch(payload)).catch(next);
  };
}
